### PR TITLE
Persist Chat Filters across zones

### LINKED ISSUE #512: Chat filters are not persistent
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened
- [x] I have checked the commit log to see if the issue has been resolved since my server was last updated
- [x] I have read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I have specified what branch this happens on branch: base

Chat filters are not permanent and revert on zone, IIRC on retail, this was actually saved per character - does this need a table in the Database? I'm happy to contribute to the project if anyone can guide me.

### LINKED ISSUE #808: Chat filters don't save after zoning
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened
- [x] I have checked the commit log to see if the issue has been resolved since my server was last updated
- [x] I have read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I have specified what branch this happens on branch: EDITME

**_Additional Information_** (Steps to reproduce/Expected behavior) **:** 
Combat log filters aren't saved after zoning and reset to stock.

### LINKED ISSUE #855: [INFO] Chat Filters / Related Data & Packets
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened
- [x] I have checked the commit log to see if the issue has been resolved since my server was last updated
- [x] I have read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I have specified what branch this happens on branch: (not a bug)

**_Additional Information_** (Steps to reproduce/Expected behavior) **:** 

I am posting the information in regards to how the client handles chat filters to keep them in-sync with the server as they are stored (partially) on the server side. _This is in the effort to see them fixed on private servers, and not rely on third-party addons/plugins to maintain things. I do not have time to work on the server side of things anymore so I am sharing this for someone else whom does have the time and wants to work on it/fix it._

## Preface

FFXI was designed with some configuration data being stored to the server when adjustments are made. Although full usage of this packet seems to be forgotten about, there are a few aspects where it has been used. That being the chat filters and the clients preferred language for finding parties. The client and server handle this configuration exchange with the following packets:

  * Client To Server: `0x00DB`
  * Server To Client: `0x00B4`

At this time, the implementation inside of LSB (and past projects) is incorrect and does not handle the data for these packets properly.

See here:

 * C2S: see here: https://github.com/LandSandBoat/server/blob/340878a6adb4fecdd9bd2e56991041928e2dcd82/src/map/packet_system.cpp#L5244
 * S2C: see here: https://github.com/LandSandBoat/server/blob/340878a6adb4fecdd9bd2e56991041928e2dcd82/src/map/packets/menu_config.cpp#L28

For the client-to-server side of things, this packet actually has multiple 'modes'. (A total of 4 are inside of the client, 2 are currently used that I can trace.)

  * 0 - Chat Filter Update
  * 1 - Preferred Language Update (Chat filters seem to be sent still anyway with this mode but wouldn't rely on it.)
  * 2 - Unknown (Potentially unused.)
  * 3 - Unknown (Potentially unused.)

## Mode 0 - Chat Filters

When the client opens the chat filter window, it is populated from the games internal configuration object manager, this uses the configuration values with the following ids:

 * 50
 * 51
 * 52
 * 53
 * 71

These hold the chat filter masks, as well as the 'hold' and 'chime' information. The server does not seem to hold onto anything other than if the filter is on or off. (It does not store hold/chime data from basic testing. The client stores this locally instead.)

When changes are made and the window is closed, the client updates the configuration table with the updated values via:

```cpp
int __thiscall sub_101923F0(int *this)
{
  FUNC_Config_SetValue_Wrap(50, this[12]);
  FUNC_Config_SetValue_Wrap(51, this[13]);
  FUNC_Config_SetValue_Wrap(52, this[14]);
  FUNC_Config_SetValue_Wrap(53, this[15]);
  return FUNC_Config_SetValue_Wrap(71, this[16]);
}
```

When this happens, a callback is triggered on the data which invokes the clients packet updater to send a configuration update:

```cpp
int __cdecl sub_101A4880(int a1, int a2)
{
  if ( a1 == 50 )
  {
    if ( !(unsigned __int8)sub_100E2930(&a1) || !a1 )
      return 1;
    *(_DWORD *)(a1 + 4) = a2;
  }
  else if ( a1 == 51 )
  {
    if ( !(unsigned __int8)sub_100E2930(&a1) || !a1 )
      return 1;
    *(_DWORD *)(a1 + 8) = a2;
  }
  else
  {
    if ( a1 != 53 || !(unsigned __int8)sub_100E2930(&a1) || !a1 )
      return 1;
    *(_BYTE *)(a1 + 1) ^= (*(_BYTE *)(a1 + 1) ^ (8 * a2)) & 0x18;
  }
  if ( !(unsigned __int8)SendConfig_Mode0_ChatFilters() )
    nullsub_12(aConfigSendFail);
  return 1;
}
```

Mode 0, as stated above is used for chat filters which is sent via:

```cpp
char __cdecl SendConfig_Mode0_ChatFilters()
{
  char result; // al

  if ( dword_104D2618 )
    result = Packet_Outgoing_0DB_CharConfig(0, (*(uint32_t*)((uint32_t)::GetModuleHandleA('FFXiMain.dll') + 0x4D2618) + 0x28A20)); // 2nd param is generally zero..
  else
    result = 0;
  return result;
}
```

The packet handler to send `0x00DB` from the client is setup like this:

```cpp
char __cdecl Packet_Outgoing_0DB_CharConfig(char mode, int param)
{
  if ( !dword_104D2618 )
    return 0;

  // Prepare the buffer for the packet..
  auto packet = FUNC_BuildOutgoingPacket(219, 1, 1);
  if ( !packet )
    return 0;

  // Populate the packet buffer data..
  const auto cfg = (*(uint32_t*)((uint32_t)::GetModuleHandleA('FFXiMain.dll') + 0x4D2618) + 0x28A00);

  *(uint32_t*)(packet + 0x04)   = 0;
  *(uint8_t*)(packet + 0x05)    = 0;
  *(uint8_t*)(packet + 0x06)    = mode; // 0 = Chat filters, 1 = Preferred Languages, 2+3 = Unknown but in the client.
  *(uint32_t*)(packet + 0x08)   = *(unit32_t*)(cfg + 0x00); // Name+Config Mask
  *(uint32_t*)(packet + 0x0C)   = *(unit32_t*)(cfg + 0x04); // Chat Filter Mask
  *(uint32_t*)(packet + 0x10)   = *(unit32_t*)(cfg + 0x08); // Chat Filter Mask
  *(uint32_t*)(packet + 0x24)   = param;

  // Mark the packet as ready for queueing/to be sent..
  FUNC_QueueOutgoingPacket_Wrap((int)packet, 40, 0);

  return 1;
}
```

## Mode 1 - Preferred Language

When the client alters the preferred languages they wish to party with, the following is called: _(Set via Menu > Party > Languages)_

```cpp
/**
 * @param {int} langs - The languages mask of what the client has enabled.
 *
 * @note
 *      These are set via Menu > Party > Languages
 */
char __cdecl SendConfig_Mode1_PreferredLanguage(int langs)
{
  return Packet_Outgoing_0DB_CharConfig(1, langs);
}
```

The preferred languages value is stored in the client configuration table as id: 152

The language mask values are:

  * 0x01 - Japanese
  * 0x02 - English
  * 0x04 - German
  * 0x08 - French
  * 0x10 - Other

## Mode 2 & 3 - Unknown

At this time, there does not seem to be references to these modes being invoked. They may have been unfinished configuration storages as the packet is larger than what is currently put to use. Mode 0 and 1 only use a small amount of the packet. There is still around 16 bytes of unused space that could be for other configuration data they had planned to sync to the server.

# Server To Client Configuration Update

When the server sends the client back their configuration information, it is sent via: `0x00B4`

```cpp
char __cdecl Packet_Incoming_0B4_CharConfig(int a1, int a2, int packet)
{
  *(_DWORD *)(a1 + 0x28A00) = *(_DWORD *)(packet + 0x04);   // Name+Config Mask
  *(_DWORD *)(a1 + 0x28A04) = *(_DWORD *)(packet + 0x08);   // Chat Filter Mask
  *(_DWORD *)(a1 + 0x28A08) = *(_DWORD *)(packet + 0x0C);   // Chat Filter Mask
  *(_WORD *)(a1 + 0x3C7EC) = *(_WORD *)(packet + 0x10);     // Linkshell Flag. (See notes below, this should be 0.)
  *(_BYTE *)(a1 + 0x28A1C) = *(_BYTE *)(packet + 0x12);     // Area Code (Used with help desk menu.)
  return 1;
}
```

This is what the linkshell breaking looks like when `0x3C7EC` is set to anything other than 0:

This is checked with the following chat handler function to prefix the linkshell messages with their given [1], [2], [3] tag:

```cpp
char *__cdecl sub_100E1A70(char *Buffer, char chatMode, char disablePrefix)
{
  char *result; // eax

  result = Buffer;
  *Buffer = 0;
  if ( !disablePrefix )
  {
    switch ( chatMode )
    {
      case 5: // Linkshell 1
        result = (char *)sprintf(Buffer, ::a1); // [1]
        break;
      case 27: // Linkshell 2
        result = (char *)sprintf(Buffer, ::a2); // [2]
        break;
      case 30: // Linkshell 3
        result = (char *)sprintf(Buffer, ::a3); // [3]
        break;
    }
  }
  return result;
}
```

The value stored in the `0x3C7EC` offset based address is used as a means to disable the linkshell prefix in the above function. Which means it appears this value should currently always be 0, and that prehaps, extended chat options were planned but never implemented. _(Disabling the prefix causes other aspects to break making this seem like an unfinished feature as seen in the pictures above.)_

The value stored in the `0x28A1C` offset based address is the client 'area code'. This appears to be used only with the help desk menu related interactions. This value is only set from the server and is generally 2. Otherwise, the client only ever defaults this value to 0 until it has received a `0x00B4` from the server with an updated code id to use. The client has a function that is used with the help desk that dumps some information that is uploaded to the server. This value is marked/named as 'AREACODE'.

If I were to take a guess, this appears to probably be a region based id/code to tell a GM what part of the world you are in. Perhaps this was used to handle telling a GM how to greet someone (ie. Good morning, Good afternoon, Good evening, etc.) or something similar. A way to test this if it is even used anymore would be to connect to retail and monitor what this value is while using a VPN. Connect from multiple different areas across the world and see if you ever get anything other than 2. (This does not seem to be tied to your client language at all.)

_This should be all the information needed to get chat filters working properly as well as to update the other incorrect data currently present for the packets involved._

_Note: Other data may be present in the packet, but the client does not expect/use it. This is strictly based on reversed information that the client is actually using._

### PR DESCRIPTION
**Pull request by zach2good**
Starting the implementation of atom0s' notes in: https://github.com/LandSandBoat/server/issues/855

Closes https://github.com/LandSandBoat/server/issues/855
Closes https://github.com/LandSandBoat/server/issues/808
Closes https://github.com/LandSandBoat/server/issues/512

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

File: src/map/packets/menu_config.cpp [hunk 1]
```diff
@@ -38,17 +38,19 @@ CMenuConfigPacket::CMenuConfigPacket(CCharEntity* PChar)
     ref<uint8>(0x06) = PChar->menuConfigFlags.byte3;
     ref<uint8>(0x07) = PChar->menuConfigFlags.byte4;
 
+    ref<uint64>(0x08) = PChar->chatFilterFlags;
```
Comment (TeoTwawki) [inline src/map/packets/menu_config.cpp:41]: 64? not 32?

Comment (zach2good) [reply src/map/packets/menu_config.cpp:41]: From atom0s's notes, it's 64-wide:
```
  *(uint32_t*)(packet + 0x0C)   = *(unit32_t*)(cfg + 0x04); // Chat Filter Mask
  *(uint32_t*)(packet + 0x10)   = *(unit32_t*)(cfg + 0x08); // Chat Filter Mask
  ```

Comment (TeoTwawki) [reply src/map/packets/menu_config.cpp:41]: kk, big damn mask

**Comment by zach2good** [comment]
Out of draft

**Review by Xaver-DaRed — APPROVED** [review]

**Pull request merged**